src/adera/SysExhaustPlume.cpp

```diff
@@ -39,52 +41,142 @@
+        auto const* name = rScene.reg_try_get<ACompName>(ent);
 
-void SysExhaustPlume::initialize_plume(ActiveScene& rScene, ActiveEnt node)
-{
-    using Magnum::GL::Mesh;
-    using Magnum::GL::Texture2D;
-    using Magnum::Trade::ImageData2D;
-    using namespace Magnum::Math::Literals;
-    using adera::shader::PlumeShader;
+        if (nullptr == name)
+        {
+            return EHierarchyTraverseStatus::Continue;
+        }
+
+        if (0 == name->m_name.compare(0, namePrefix.size(), namePrefix))
+        {
+            plumeNode = ent;
+            return EHierarchyTraverseStatus::Stop;  // terminate search
+        }
+        return EHierarchyTraverseStatus::Continue;
+    };
+
+    SysHierarchy::traverse(rScene, part, findPlumeHandle);
+
+    if (plumeNode == entt::null)
+    {
+        SPDLOG_LOGGER_ERROR(
+                rScene.get_application().get_logger(),
+                "ERROR: could not find plume anchor for MachineRocket {}",
+                part);
+        return;
+    }
+
+    SPDLOG_LOGGER_INFO(
+            rScene.get_application().get_logger(),
+            "MachineRocket {}'s associated plume: {}",
+            part, plumeNode);
+
+    // Get plume effect
+
+    using osp::DependRes;
+    using osp::Package;
 
-    auto const& plume = rScene.reg_get<ACompExhaustPlume>(node);
+    Package &rPkg = rScene.get_application().debug_find_package("lzdb");
 
-    if (rScene.get_registry().all_of<ACompShader>(node))
+    std::string_view effectName = rScene.reg_get<ACompName>(plumeNode).m_name;
+    effectName.remove_prefix(3); // removes "fx_"
+
+    DependRes<PlumeEffectData> plumeEffect = rPkg.get<PlumeEffectData>(effectName);
+    if (plumeEffect.empty())
     {
-        return;  // plume already initialized
+        SPDLOG_LOGGER_ERROR(rScene.get_application().get_logger(),
+                            "ERROR: couldn't find plume effect  {}", effectName);
+        return;
     }
 
-    DependRes<PlumeEffectData> plumeEffect = plume.m_effect;
+    rScene.reg_emplace<ACompExhaustPlume>(plumeNode, mach, plumeEffect);
+
+    // Add Plume mesh
+
+    using osp::AssetImporter;
+    using Magnum::GL::Mesh;
+
+    using osp::active::ACompMesh;
+    using osp::active::ACompTransparent;
+    using osp::active::ACompVisible;
 
-    Package& pkg = rScene.get_application().debug_find_package("lzdb");
+    Package &rGlResources = rScene.get_context_resources();
+    DependRes<Mesh> plumeMesh = rGlResources.get<Mesh>(plumeEffect->m_meshName);
 
-    // Get plume mesh
-    Package& glResources = rScene.get_context_resources();
-    DependRes<Mesh> plumeMesh = glResources.get<Mesh>(plumeEffect->m_meshName);
+    // Compile the plume mesh if not done so yet
     if (plumeMesh.empty())
     {
-        plumeMesh = AssetImporter::compile_mesh(plumeEffect->m_meshName, pkg, glResources);
+        plumeMesh = AssetImporter::compile_mesh(
+                        plumeEffect->m_meshName, rPkg, rGlResources);
     }
```

Comment:
I seem to remember writing `compile_mesh()` for the specific purpose of simplifying this process, but since this get-or-compile process will probably be used every time a resource is fetched like this, it might be worth writing a specific function called `get_or_load` or similar that basically just does


```cpp
template<typename T>
DependRes<T> get_or_load(std::string_view resName)
{
    Package &rPkg = rScene.get_application().debug_find_package("lzdb");
    Package& rGlResources = rScene.get_context_resources();
    DependRes<T> res = rGlResources.get<T>(resName);
    if (res.empty())
    {
        res = AssetImporter::compile_resource<T>(resName, rPkg, rGlResources);
    }
    return res;
}
```

Not strictly pertinent to this PR, but could be a good time to do it.

Also, we should probably move away from the name `rGlResources` and use `rGfxResources` or similar since presumably we will replace the GL backend in the future or offer multiple choices, so we should aim to be as API agnostic as possible (although it's just a name so obviously it's not a big deal).

Reply:
I see that you used `rCtxResources` elsewhere in which case carry on

Reply:
This is likely to be  moved into a dedicated system to decouple gl-specific functions from systems that create meshes, and better control which thread they are run on.